### DISCUSSION TITLE
Qemu integration with the test infrastructure

### DISCUSSION THREAD

**Discussion by mbait** [modm-io/modm#1081] [Ideas]
Qemu supports AVR instruction set along with UART and 16-bit timer emulation. Does it make sense to put some effort into integrating it with the test system? For example, the fiber example for mega2560 is emulated just fine on Qemu. I see two problems though. One - the clock signal is not emulated realistically, so the fiber example gives weird numbers. Two - there's no easy way to shutdown a qemu instance other than emitting a halt instruction of the target platform, but there's no such instruction in AVR. I'd be happy to hear your thoughts.

**Comment by salkinium** [comment]
Yes, especially if you can attach GDB to AVR on Qemu. I still don't have a working JTAG setup for debugging AVR, which was quite annoying when my fiber implementation did not work and I had no way to debug it.
You would create a [new target here](https://github.com/modm-io/modm/tree/develop/tools/devices), maybe something like `qemu/avr-atmega2560.xml` with the same platform/family but different peripheral type names.
Then you can switch the peripheral selection and apply workaround for things like the cycle counter.

We can probably SIGTERM it externally by wrapping Qemu in Python, we do something similar for the debug probes.

There's also this: https://github.com/buserror/simavr with GDB support. But I don't know how it compares.
